// @sap/cds/libx/_runtime/sqlite/customBuilder/index.js
class ReferenceBuilder extends require('../../db/sql-builder').ReferenceBuilder {}
const ExpressionBuilder = require('./CustomExpressionBuilder')
const FunctionBuilder = require('./CustomFunctionBuilder')
const SelectBuilder = require('./CustomSelectBuilder')
const UpsertBuilder = require('./CustomUpsertBuilder')
class UpdateBuilder extends require('../../db/sql-builder').UpdateBuilder {}
class DeleteBuilder extends require('../../db/sql-builder').DeleteBuilder {}

module.exports = {
  ReferenceBuilder: extend(ReferenceBuilder).with({
    FunctionBuilder
  }),
  ExpressionBuilder: extend(ExpressionBuilder).with({
    ReferenceBuilder,
    SelectBuilder,
    FunctionBuilder
  }),
  FunctionBuilder: extend(FunctionBuilder).with({
    ExpressionBuilder,
    ReferenceBuilder,
    SelectBuilder
  }),
  SelectBuilder: extend(SelectBuilder).with({
    ExpressionBuilder,
    ReferenceBuilder,
    FunctionBuilder,
    SelectBuilder
  }),
  UpsertBuilder,
  UpdateBuilder: extend(UpdateBuilder).with({
    ReferenceBuilder,
    ExpressionBuilder
  }),
  DeleteBuilder: extend(DeleteBuilder).with({
    ReferenceBuilder,
    ExpressionBuilder
  })
}

function extend(clazz) {
  return {
    with(properties) {
      for (let p in properties) {
        Object.defineProperty(clazz.prototype, p, { value: properties[p] })
      }
      return clazz
    }
  }
}
