// @sap/cds/lib/utils/colors.js
const none = process.stdout.isTTY && !process.env.NO_COLOR || process.env.FORCE_COLOR ? null : ''
const colors = {
  RESET:        none ?? '\x1b[0m',
  BOLD:         none ?? '\x1b[1m',
  BRIGHT:       none ?? '\x1b[1m',
  DIMMED:       none ?? '\x1b[2m',
  ITALIC:       none ?? '\x1b[3m',
  UNDER:        none ?? '\x1b[4m',
  BLINK:        none ?? '\x1b[5m',
  FLASH:        none ?? '\x1b[6m',
  INVERT:       none ?? '\x1b[7m',
  BLACK:        none ?? '\x1b[30m',
  RED:          none ?? '\x1b[31m',
  GREEN:        none ?? '\x1b[32m',
  YELLOW:       none ?? '\x1b[33m',
  BLUE:         none ?? '\x1b[34m',
  PINK:         none ?? '\x1b[35m',
  CYAN:         none ?? '\x1b[36m',
  LIGHT_GRAY:   none ?? '\x1b[37m',
  DEFAULT:      none ?? '\x1b[39m',
  GRAY:         none ?? '\x1b[90m',
  LIGHT_RED:    none ?? '\x1b[91m',
  LIGHT_GREEN:  none ?? '\x1b[92m',
  LIGHT_YELLOW: none ?? '\x1b[93m',
  LIGHT_BLUE:   none ?? '\x1b[94m',
  LIGHT_PINK:   none ?? '\x1b[95m',
  LIGHT_CYAN:   none ?? '\x1b[96m',
  WHITE:        none ?? '\x1b[97m',
  bg: {
    BLACK: none ?? '\x1b[40m',
    RED: none ?? '\x1b[41m',
    GREEN: none ?? '\x1b[42m',
    YELLOW: none ?? '\x1b[43m',
    BLUE: none ?? '\x1b[44m',
    PINK: none ?? '\x1b[45m',
    CYAN: none ?? '\x1b[46m',
    WHITE: none ?? '\x1b[47m',
    DEFAULT: none ?? '\x1b[49m',
    LIGHT_GRAY: none ?? '\x1b[100m',
    LIGHT_RED: none ?? '\x1b[101m',
    LIGHT_GREEN: none ?? '\x1b[102m',
    LIGHT_YELLOW: none ?? '\x1b[103m',
    LIGHT_BLUE: none ?? '\x1b[104m',
    LIGHT_PINK: none ?? '\x1b[105m',
    LIGHT_CYAN: none ?? '\x1b[106m',
    LIGHT_WHITE: none ?? '\x1b[107m',
  },
}
module.exports = colors
